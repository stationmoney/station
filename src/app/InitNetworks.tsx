import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
