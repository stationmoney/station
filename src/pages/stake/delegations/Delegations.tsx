import { useTranslation } from "react-i18next"
import { AccAddress, Coin, Delegation } from "@terra-money/feather.js"
import { getMaxHeightStyle } from "utils/style"
import { combineState } from "data/query"
import { useMemoizedPrices } from "data/queries/coingecko"
import { useInterchainDelegations } from "data/queries/staking"
import { ValidatorLink } from "components/general"
import { ModalButton } from "components/feedback"
import { Table } from "components/layout"
import { Read } from "components/token"
import StakedCard from "../components/StakedCard"
import { useNativeDenoms } from "data/token"
import styles from "../CardModal.module.scss"

const Delegations = () => {
  const { t } = useTranslation()
  const readNativeDenom = useNativeDenoms()
  const { data: prices, ...pricesState } = useMemoizedPrices()

  const interchainDelegations = useInterchainDelegations()

  const delegations: Delegation[] = interchainDelegations.reduce(
    (acc, { data }) => (data ? [...data?.delegation, ...acc] : acc),
    [] as Delegation[]
  )
  const state = combineState(pricesState, ...interchainDelegations)

  /* render */
  const title = t("Delegations")

  const render = () => {
    if (!delegations || !prices) return null

    const total = delegations.reduce((acc, { balance }) => {
      const { token, decimals } = readNativeDenom(balance.denom)
      return (
        acc +
        (balance.amount.toNumber() * (prices[token]?.price || 0)) /
          10 ** decimals
      )
    }, 0)

    return (
      <ModalButton
        title={title}
        renderButton={(open) => (
          <StakedCard
            {...state}
            title={
              <div className={styles.header_wrapper}>
                {title}
                {total !== -1 && (
                  <span className={styles.view_more}>View More</span>
                )}
              </div>
            }
            value={total?.toString() || "0"}
            onClick={open}
            cardName={"delegations"}
          />
        )}
      >
        <Table
          dataSource={delegations}
          sorter={({ balance: { amount: a } }, { balance: { amount: b } }) =>
            b.minus(a).toNumber()
          }
          columns={[
            {
              title: t("Validator"),
              dataIndex: "validator_address",
              render: (address: AccAddress) => (
                <ValidatorLink address={address} internal />
              ),
            },
            {
              title: t("Delegated"),
              dataIndex: "balance",
              render: (balance: Coin) => <Read {...balance.toData()} />,
              align: "right",
            },
          ]}
          style={getMaxHeightStyle(320)}
        />
      </ModalButton>
    )
  }

  return render()
}

export default Delegations
