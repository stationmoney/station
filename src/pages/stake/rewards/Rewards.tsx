import { useTranslation } from "react-i18next"
import { useCurrency } from "data/settings/Currency"
import { combineState } from "data/query"
import { calcRewardsValues, useRewards } from "data/queries/distribution"
import { useExchangeRates, useMemoizedPrices } from "data/queries/coingecko"
import { useMemoizedCalcValue } from "data/queries/coingecko"
import { useNativeDenoms, WithTokenItem } from "data/token"
import { ModalButton } from "components/feedback"
import { TokenCard, TokenCardGrid } from "components/token"
import { TooltipIcon } from "components/display"
import StakedCard from "../components/StakedCard"
import RewardsTooltip from "../RewardsTooltip"
import styles from "../CardModal.module.scss"
import { Coin } from "@terra-money/feather.js"

const Rewards = () => {
  const { t } = useTranslation()
  const currency = useCurrency()
  const calcValue = useMemoizedCalcValue()
  const readNativeDenom = useNativeDenoms()

  const { data: prices, ...pricesState } = useMemoizedPrices()
  const { data: rewards, ...rewardsState } = useRewards()
  const { data: exchangeRates, ...exchangeRatesState } = useExchangeRates()

  const state = combineState(rewardsState, exchangeRatesState, pricesState)

  /* render */
  const title = t("Staking rewards")
  const render = () => {
    let sameDenom: boolean = true
    const coinsValue =
      !rewards || !prices
        ? 0
        : rewards.total.toData()?.reduce((acc, { amount, denom }) => {
            const { token, decimals } = readNativeDenom(denom)
            if (denom !== rewards?.total.toData()[0].denom) {
              sameDenom = false
            }
            return (
              acc +
              (parseInt(amount) * (prices?.[token]?.price || 0)) /
                10 ** decimals
            )
          }, 0)

    const totalToDisplay = coinsValue

    let list: Coin.Data[] = []
    if (rewards) {
      const { total } = calcRewardsValues(rewards, currency.id, calcValue)
      list = total.list
    }

    const showTokens = totalToDisplay === -1 || sameDenom

    console.log({ totalToDisplay, sameDenom, showTokens })

    return (
      <ModalButton
        title={title}
        renderButton={(open) => (
          <StakedCard
            {...state}
            title={
              <div className={styles.header_wrapper}>
                <TooltipIcon content={<RewardsTooltip />} placement="bottom">
                  {title}
                </TooltipIcon>
                {!showTokens && (
                  <span className={styles.view_more}>View More</span>
                )}
              </div>
            }
            value={totalToDisplay?.toString() || "-1"}
            showTokens={showTokens}
            onClick={open}
            cardName="rewards"
          ></StakedCard>
        )}
      >
        <TokenCardGrid maxHeight>
          {list?.map(({ amount, denom }) => (
            <WithTokenItem token={denom} key={denom}>
              {(item) => (
                <TokenCard
                  {...item}
                  amount={amount}
                  value={calcValue({ amount, denom })}
                />
              )}
            </WithTokenItem>
          ))}
        </TokenCardGrid>
      </ModalButton>
    )
  }

  return render()
}

export default Rewards
