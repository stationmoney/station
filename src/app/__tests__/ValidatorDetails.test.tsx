import React from "react"
import ValidatorDetails from "pages/stake/ValidatorDetails"
import { RecoilRoot } from "recoil"
import { NetworksProvider } from "../InitNetworks"
import { render } from "@testing-library/react"
import { mockNetworks } from "./__mocks__/Networks.mock"
import { mockValidatorDetails } from "./__mocks__/ValidatorDetails.mock"
import { MemoryRouter } from "react-router-dom"
import { mockDelegation, mockDelegations } from "./__mocks__/Delegations.mock"
import { mockRewards } from "./__mocks__/Rewards.mock"
import { mockExchangeRates } from "./__mocks__/ExchangeRates.mock"
import { mockWhitelist } from "./__mocks__/Whitelist.mock"

function renderComponent() {
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = jest.fn()

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <MemoryRouter>
        <RecoilRoot>
          <ValidatorDetails />
        </RecoilRoot>
      </MemoryRouter>
    </NetworksProvider>
  )
}

jest.mock("lottie-react", () => ({
  Lottie: jest.fn(),
}))

jest.mock("../../pages/stake/useAddressParams", () => ({
  __esModule: true,
  default: () => "terravaloper1fzx6z3qlwjyjjt3e5q6sakvt74n08uq9alzae3",
}))

jest.mock("../../data/wallet", () => {
  const mockUseNetwork = () => {
    return mockNetworks
  }

  const mockUseNetworkName = () => {
    return "testnet"
  }

  return {
    useNetwork: mockUseNetwork,
    useNetworkName: mockUseNetworkName,
  }
})

jest.mock("../../data/queries/staking", () => {
  const mockUseValidator = () => mockValidatorDetails
  const mockUseDelegation = () => mockDelegation
  const mockUseDelegations = () => mockDelegations

  return {
    useValidator: mockUseValidator,
    useDelegation: mockUseDelegation,
    useDelegations: mockUseDelegations,
  }
})

jest.mock("../../data/settings/Theme", () => {
  const mockUseThemeFavicon = () => "/static/media/favicon.1e08d51d.svg"

  return {
    useThemeFavicon: mockUseThemeFavicon,
  }
})

jest.mock("../../data/queries/distribution", () => {
  const mockUseRewards = () => {
    return mockRewards
  }

  return {
    useRewards: mockUseRewards,
  }
})

jest.mock("../../data/queries/coingecko", () => {
  const mockUseExchangeRates = () => {
    return mockExchangeRates
  }

  const mockUseMemoizedCalcValue = () => {
    return (
      ({ amount, denom }) => {
        return Number(amount) * Number(mockExchangeRates[denom] ?? 0)
      },
      [mockExchangeRates]
    )
  }

  return {
    useExchangeRates: mockUseExchangeRates,
    useMemoizedCalcValue: mockUseMemoizedCalcValue,
  }
})

jest.mock("../../data/queries/chains", () => {
  const actual = jest.requireActual("../../data/queries/chains")

  const mockUseWhitelist = () => {
    return mockWhitelist
  }

  return {
    ...actual,
    useWhitelist: mockUseWhitelist,
  }
})

describe("Links component matches snapshots", () => {
  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
