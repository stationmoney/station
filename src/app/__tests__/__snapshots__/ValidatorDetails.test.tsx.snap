// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links component matches snapshots matches original component 1`] = `
<DocumentFragment>
  <article
    class="page"
  >
    <div
      class="container grid"
    >
      <header
        class="header"
      >
        <div
          class="titleWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
          <h1
            class="title"
          >
            Validator details
          </h1>
        </div>
      </header>
      <section
        class="main"
      >
        <div
          class="row start"
        >
          <div
            class="col"
            style="flex: 3;"
          >
            <div
              class="col"
              style="flex: 1;"
            >
              <article
                class="card default"
              >
                <section
                  class="main"
                >
                  <div
                    class="grid"
                    style="gap: 16px;"
                  >
                    <header
                      class="header"
                    >
                      <img
                        alt=""
                        class="icon"
                        height="60"
                        src="https://raw.githubusercontent.com/terra-money/validator-images/main/images/0A6AF02D1557E5B4.jpg"
                        width="60"
                      />
                      <div
                        class="grid"
                        style="gap: 4px;"
                      >
                        <div
                          class="flex start"
                          style="gap: 10px;"
                        >
                          <h1
                            class="moniker"
                          >
                            polkachu.com
                          </h1>
                          <span
                            class="inline tag bg-success"
                          >
                            Active
                          </span>
                        </div>
                        <div
                          class="flex start link"
                          style="gap: 4px;"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                            data-testid="LaptopOutlinedIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2H0v2h24v-2h-4zM4 6h16v10H4V6z"
                            />
                          </svg>
                          <a
                            href="https://polkachu.com"
                            rel="noopener noreferrer"
                            target="_blank"
                          >
                            https://polkachu.com
                          </a>
                        </div>
                      </div>
                    </header>
                    <p>
                      Polkachu is the trusted staking service provider for blockchain projects. 100% refund for downtime slash. Contact us at hello@polkachu.com
                    </p>
                  </div>
                </section>
              </article>
              <article
                class="card bordered grid"
              >
                <header
                  class="header"
                >
                  <h1
                    class="title"
                  >
                    Commission
                  </h1>
                  <div
                    class="flex extra"
                  />
                </header>
                <section
                  class="main"
                >
                  <div
                    class="grid numbers"
                  >
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Current
                      </h1>
                      <p
                        class="value"
                      >
                        5.00%
                      </p>
                    </div>
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Max
                      </h1>
                      <p
                        class="value"
                      >
                        10.00%
                      </p>
                    </div>
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Max daily change
                      </h1>
                      <p
                        class="value"
                      >
                        5.00%
                      </p>
                    </div>
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Last changed
                      </h1>
                      <p
                        class="value"
                      >
                        <span>
                          11 months ago
                        </span>
                      </p>
                    </div>
                  </div>
                </section>
              </article>
              <article
                class="card default"
              >
                <section
                  class="main"
                >
                  <div
                    class="grid addresses"
                    style="gap: 12px;"
                  >
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Operator address
                      </h1>
                      <p
                        class="value"
                      >
                        <a
                          class="link"
                          href="https://terrascope.info/testnet/validator/terravaloper1fzx6z3qlwjyjjt3e5q6sakvt74n08uq9alzae3"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          terravaloper1fzx6z3qlwjyjjt3e5q6sakvt74n08uq9alzae3
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                            data-testid="CallMadeIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M9 5v2h6.59L4 18.59 5.41 20 17 8.41V15h2V5z"
                            />
                          </svg>
                        </a>
                      </p>
                    </div>
                    <div
                      class="grid item"
                    >
                      <h1
                        class="title"
                      >
                        Wallet address
                      </h1>
                      <p
                        class="value"
                      >
                        <a
                          class="link"
                          href="https://terrascope.info/testnet/address/terra1fzx6z3qlwjyjjt3e5q6sakvt74n08uq9aswqfz"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          terra1fzx6z3qlwjyjjt3e5q6sakvt74n08uq9aswqfz
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                            data-testid="CallMadeIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M9 5v2h6.59L4 18.59 5.41 20 17 8.41V15h2V5z"
                            />
                          </svg>
                        </a>
                      </p>
                    </div>
                  </div>
                </section>
              </article>
            </div>
          </div>
          <div
            class="col"
            style="flex: 2;"
          >
            <div
              class="col"
              style="flex: 1;"
            >
              <article
                class="card default grid"
              >
                <header
                  class="header"
                >
                  <h1
                    class="title"
                  >
                    My delegations
                  </h1>
                  <div
                    class="flex extra"
                  />
                </header>
                <section
                  class="main"
                >
                  <div
                    class="grid"
                    style="gap: 20px;"
                  >
                    <section>
                      <span
                        class="component block total"
                      >
                        1
                        <span
                          class="small"
                        >
                          .00
                        </span>
                        <span
                          class="small"
                        >
                           LUNA
                        </span>
                      </span>
                    </section>
                  </div>
                </section>
              </article>
              <article
                class="card default grid"
              >
                <header
                  class="header"
                >
                  <h1
                    class="title"
                  >
                    My rewards
                  </h1>
                  <div
                    class="flex extra"
                  />
                </header>
                <section
                  class="main"
                >
                  <div
                    class="grid"
                    style="gap: 20px;"
                  >
                    <section>
                      <span
                        class="component total"
                      >
                        0
                        <span
                          class="small"
                        >
                          .00
                        </span>
                        <span
                          class="small"
                        >
                           LUNA
                        </span>
                      </span>
                       
                      <span
                        class="small"
                      />
                    </section>
                    <div
                      class="actions stretch"
                    >
                      <span
                        class="button default block disabled"
                        disabled=""
                        to="/rewards"
                      >
                        Withdraw rewards
                      </span>
                    </div>
                  </div>
                </section>
              </article>
            </div>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
